Mutable pod scheduling directives

danielvegamyhre opened the pull request:
#### What type of PR is this?
/kind feature
/sig scheduling

Optionally add one or more of the following kinds if applicable:
/kind api-change


#### What this PR does / why we need it:
See [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3838-pod-mutable-scheduling-directives) for context.

#### Which issue(s) this PR fixes:
Fixes #116160

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Relax API validation to allow pod node selector to be mutable for gated pods (additions only, no deletions or mutations).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
[KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3838-pod-mutable-scheduling-directives)
```

danielvegamyhre committed:
mutable pod scheduling directives

ahg-g commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4668,6 +4670,14 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 
+	// Only additions to spec.nodeSelector are allowed (no deletions or mutations).
+	// If it is empty, is it allowed to be set.
+	podIsGated := len(oldPod.Spec.SchedulingGates) > 0 && len(mungedPodSpec.SchedulingGates) > 0
```
also, we want to allow this only when `PodSchedulingReadiness` feature gate is enabled, so you probably need to do this in PodValidationOptions

ahg-g commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -12485,6 +12485,70 @@ func TestValidatePodUpdate(t *testing.T) {
 			err:  "",
 			test: "update pod spec schedulingGates: legal deletion",
 		},
+		{
```
add a test where the old has a gate, but the new doesn't; in this case we should allow adding a new nodeSelector

ahg-g reviewed (commented):
This is in the right track; please note that we need to do this for nodeAffinity too.

ahg-g commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4668,6 +4674,22 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+		if nodeAffinityHasMutated {
+			allErrs = append(allErrs, validateNodeAffinityMutation(specPath.Child("affinity").Child("nodeAffinity"), oldPod.Spec.Affinity.NodeAffinity, mungedPodSpec.Affinity.NodeAffinity)...)
+		}
+		return allErrs
```
we should have a test where we are mutating something else after this clause to ensure that the deepequal below is hit and causes an error. I guess a test that adds an item to a non-empty NodeSelectorTerms list should cover this path.

ahg-g commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7274,75 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+	return allErrs
+}
+
+func validateNodeAffinityMutation(nodeAffinityPath *field.Path, oldNodeAffinity, newNodeAffinity *core.NodeAffinity) field.ErrorList {
```
we should validate that the number of old and new NodeSelectorTerms are the same, no additions to a non-zero length NodeSelectorTerms is allowed because semantically it is an OR.

ahg-g commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -12485,6 +12486,574 @@ func TestValidatePodUpdate(t *testing.T) {
+			opts: PodValidationOptions{
+				AllowMutableNodeSelectorAndNodeAffinity: true,
+			},
+			test: "adding node selector is allowed for gated pods",
```
add a similar test where the old is gated, but the new is not. It should pass.

danielvegamyhre replied:
There is one already, it is called `"old pod spec has scheduling gate, new pod spec does not, and node selector is added"`.

ahg-g commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -12485,6 +12486,574 @@ func TestValidatePodUpdate(t *testing.T) {
+			err:  "spec.nodeSelector: Invalid value:",
+			test: "modifying value of existing node selector is not allowed",
+		},
+		{
```
add a test where a new item is added to `NodeSelectorTerms` and the pod is gated, it should fail

danielvegamyhre commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4668,6 +4674,22 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 
+	// Additions to spec.nodeSelector are allowed (no deletions or mutations) for gated pods.
+	podIsGated := len(oldPod.Spec.SchedulingGates) > 0
+	nodeSelectorHasMutated := !apiequality.Semantic.DeepEqual(oldPod.Spec.NodeSelector, mungedPodSpec.NodeSelector)
```
Thanks for the suggestion, I have added it with a small change to prevent nil deference errors when spec.Affinity is nil.

danielvegamyhre commented:
@ahg-g Thanks for the review, I have addressed your comments and this is ready for another look. I will also start working on integration tests.

ahg-g commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7271,82 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+	// NodeSelectorTerms are semantically OR-ed, so adding a new one would
+	// broaden the constraint rather than limit it further.
+	if len(oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) != len(newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) {
+		return append(allErrs, field.Invalid(nodeAffinityPath.Child("requiredDuringSchedulingIgnoredDuringExecution").Child("nodeSelectorTerms"), newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms, "no additions to non-zero length NodeSelectorTerms are allowed"))
```
```suggestion
		return append(allErrs, field.Invalid(nodeAffinityPath.Child("requiredDuringSchedulingIgnoredDuringExecution").Child("nodeSelectorTerms"), newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms, "no additions to non-empty NodeSelectorTerms list are allowed"))
```

ahg-g reviewed (commented):
Thanks, lets add integration tests.

ahg-g commented:
/label api-review 

@liggitt this is ready

danielvegamyhre commented:
@ahg-g Just added integration tests, let me know if you have any feedback.

danielvegamyhre commented:
/retest

ahg-g commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4668,6 +4674,23 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 	}
 
+	podIsGated := len(oldPod.Spec.SchedulingGates) > 0
+	if opts.AllowMutableNodeSelectorAndNodeAffinity && podIsGated {
```
All updates to `PreferredDuringSchedulingIgnoredDuringExecution` are allowed too, I missed to add a comment on that.

danielvegamyhre replied:
Added validation logic allowing any updates to PreferredDuringSchedulingIgnoredDuringExecution and a few test cases for this.

danielvegamyhre commented:
/retest

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4668,6 +4674,37 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+			mungedPodSpec.NodeSelector = oldPod.Spec.NodeSelector // +k8s:verify-mutation:reason=clone
+		}
+
+		if oldPod.Spec.Affinity != nil && mungedPodSpec.Affinity != nil {
```
it's hard to follow the munging/validation being conditional on whether the new and old top-level affinity fields are nil, when those _also_ contain pod affinity/anti-affinity which are not allowed to mutate

A structure like this might be easier to follow:

```go
var oldNodeAffinity *core.NodeAffinity
if oldPod.Spec.Affinity != nil {
  oldNodeAffinity = oldPod.Spec.Affinity.NodeAffinity
}

var mungedNodeAffinity *core.NodeAffinity
if mungedPod.Spec.Affinity != nil {
  mungedNodeAffinity = mungedPod.Spec.Affinity.NodeAffinity
}

if !apiequality.Semantic.DeepEqual(oldNodeAffinity, mungedNodeAffinity) {
  allErrs = append(allErrs, validateNodeAffinityMutation(specPath.Child("affinity").Child("nodeAffinity"), mungedNodeAffinity, oldNodeAffinity)...)
  if mungedPod.Spec.Affinity != nil {
    mungedPod.Spec.Affinity.NodeAffinity = oldNodeAffinity // +k8s:verify-mutation:reason=clone
  }  
}
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
 	return allErrs
 }
+
+func validateNodeSelectorMutation(fldPath *field.Path, oldNodeSelector, newNodeSelector map[string]string) field.ErrorList {
```
```suggestion
func validateNodeSelectorMutation(fldPath *field.Path, newNodeSelector, oldNodeSelector map[string]string) field.ErrorList {
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+	return allErrs
+}
+
+func validateNodeAffinityMutation(nodeAffinityPath *field.Path, oldNodeAffinity, newNodeAffinity *core.NodeAffinity) field.ErrorList {
```
```suggestion
func validateNodeAffinityMutation(nodeAffinityPath *field.Path, newNodeAffinity, oldNodeAffinity *core.NodeAffinity) field.ErrorList {
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+
+	// RequiredDuringSchedulingIgnoredDuringExecution must not be nil for both old and new nodeAffinity
+	// in order to perform this validation.
+	if oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil || newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
```
if `oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution` is non-nil, `newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution` being `nil` should not short-circuit the rest of this function, we should continue and that should be an error, right?

If so, add a unit test that would have caught this incorrect short-circuit, and simplify to this:

```go
if oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
  return
}
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+
+	// NodeSelectorTerms are semantically OR-ed, so adding a new one would
+	// broaden the constraint rather than limit it further.
+	if len(oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) != len(newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) {
```
> adding a new one would broaden the constraint rather than limit it further

this isn't correct if len(old) == 0 and len(new) > 0, right? I thought that was allowed

danielvegamyhre replied:
Correct, fixed this.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+	if oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil || newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		return allErrs
+	}
+
```
since `newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution` could be nil, and that should be an error, a similar pattern of calculating the effective node selector terms and then comparing them could work even when parent fields are nil:

```go
oldTerms := oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
var newTerms []core.NodeSelectorTerm
if newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
  newTerms = newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
}

if len(oldTerms) != len(newTerms) {
  ...
}
...
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+	// to matchExpressions or fieldExpressions are allowed.
+	for _, oldTerm := range oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		found := false
+		for _, newTerm := range newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
```
This lets oldTerms change locations in the list... is there a reason for that flexibility?

I expected something more like:
```go
for i := range oldTerms {
  if !validateNodeSelectorTermHasOnlyAdditions(newTerms[i], oldTerms[i]) {
     ... error
  }
}
```

liggitt replied:
yes, order of lists is deterministic (previously we required DeepEqual to return true on these since they were completely immutable)

ahg-g replied:
Sounds good to me too; I also thought about suggesting this to simplify the code, but the KEP didn't discuss that the order should continue to be deterministic...

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+	return allErrs
+}
+
+func validateNodeSelectorTermHasOnlyAdditions(oldTerm, newTerm core.NodeSelectorTerm) bool {
```
```suggestion
func validateNodeSelectorTermHasOnlyAdditions(newTerm, oldTerm core.NodeSelectorTerm) bool {
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+
+func validateNodeSelectorTermHasOnlyAdditions(oldTerm, newTerm core.NodeSelectorTerm) bool {
+	// Check MatchExpressions.
+	for _, oldExpr := range oldTerm.MatchExpressions {
+		found := false
+		for _, newExpr := range newTerm.MatchExpressions {
```
this lets old expressions move location... is there a reason for this flexibility? I expected:

```go
if l := len(oldTerm.MatchExpressions); l > 0 {
  if len(newTerm.MatchExpressions) < l {
    return false
  }
  if !reflect.DeepEqual(newTerm.MatchExpressions[:l-1], oldTerm.MatchExpressions) {
    return false
  }
}
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+		}
+	}
+
+	// Check MatchFields.
```
can this simplify to:

```go
if l := len(oldTerm.MatchFields); l > 0 {
  if len(newTerm.MatchFields) < l {
    return false
  }
  if !reflect.DeepEqual(newTerm.MatchFields[:l-1], oldTerm.MatchFields) {
    return false
  }
}
```

liggitt commented on test/integration/pods/pods_test.go:
```diff
@@ -674,3 +677,928 @@ func TestPodUpdateEphemeralContainers(t *testing.T) {
+		update                *v1.Pod
+		enableSchedulingGates bool
+		err                   string
+	}{
```
agree, pretty much all of these can be unit tests

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
+func validateNodeSelectorMutation(fldPath *field.Path, oldNodeSelector, newNodeSelector map[string]string) field.ErrorList {
+	var allErrs field.ErrorList
+	// If old node selector was nil, anything can be set.
+	if oldNodeSelector == nil {
```
```suggestion
	if len(oldNodeSelector) == 0 {
```

danielvegamyhre commented:
@liggitt Thanks for the review, I've addressed your comments and included your suggestions (or something similar). Let me know if you have any additional feedback.

danielvegamyhre commented:
/retest

valaparthvi commented:
Hi @danielvegamyhre. I am from the Bug triage shadow team for K8s release 1.27. I am checking in to make sure this PR is on track for 1.27 release.

Please note that Code Freeze begins on March 15, post this date, only critical and release-blocker issues/PRs will be accepted.

danielvegamyhre commented:
@valaparthvi Yes this is on track to be submitted before March 15th so it should be included in v1.27.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4671,6 +4673,49 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+		// Validate node affinity mutations.
+		var oldNodeAffinity *core.NodeAffinity
+		if oldPod.Spec.Affinity != nil {
+			oldNodeAffinity = oldPod.Spec.Affinity.NodeAffinity
```
yeah, that's fine... the verify job flags things that look like they might be mutating input objects in validation (they aren't in this case, they're working on a copy)

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4671,6 +4673,49 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+			}
+		}
+
+		// All updates to PreferredDuringSchedulingIgnoredDuringExecution are allowed for gated pods.
+		var oldPreferredSchedulingTerms []core.PreferredSchedulingTerm
+		if oldNodeAffinity != nil {
+			oldPreferredSchedulingTerms = oldNodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution
+		}
+		if len(oldPreferredSchedulingTerms) != 0 {
+			if mungedPodSpec.Affinity == nil {
+				mungedPodSpec.Affinity = &core.Affinity{
+					NodeAffinity: &core.NodeAffinity{},
+				}
+			}
+			mungedPodSpec.Affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution = oldPreferredSchedulingTerms // +k8s:verify-mutation:reason=clone
+		}
```
none of this should be needed, since we set `mungedPodSpec.Affinity.NodeAffinity = oldNodeAffinity` above, right?

danielvegamyhre replied:
Ah, I was using this to handle the test case `preferredDuringSchedulingIgnoredDuringExecution can have removals for gated pods`, however, now that you mention it, it _should_ have been handled by the line `mungedPodSpec.Affinity.NodeAffinity = oldNodeAffinity` above, however, that line was only executed when `mungedPodSpec.Affinity != nil`, and in this test case the new pod update had the affinity field as nil, so the line wasn't executed. 

I updated this to also still do `mungedPodSpec.Affinity.NodeAffinity = oldNodeAffinity` even if new affinity is nil, removed the highlighted code above, and all tests are passing.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7297,3 +7341,71 @@ func validatePvNodeAffinity(newPvNodeAffinity, oldPvNodeAffinity *core.VolumeNod
+
+	oldTerms := oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
+	var newTerms []core.NodeSelectorTerm
+	if newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
```
newNodeAffinity can be nil, and this access will panic, add a unit test that covers this case

danielvegamyhre replied:
Added a test case for this

danielvegamyhre commented:
/retest

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4671,6 +4673,37 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+		if !apiequality.Semantic.DeepEqual(oldNodeAffinity, mungedNodeAffinity) {
+			allErrs = append(allErrs, validateNodeAffinityMutation(specPath.Child("affinity").Child("nodeAffinity"), mungedNodeAffinity, oldNodeAffinity)...)
+			if mungedPodSpec.Affinity == nil {
+				mungedPodSpec.Affinity = &core.Affinity{
```
Unconditionally setting `mungedPodSpec.Affinity` to a non-nil value here *may* be correct here because of comparisons made earlier but it's a little hard to reason about.

Would it be clearer to only switch mungedPodSpec.Affinity from nil to non-nil if needed to copy over a non-nil oldNodeAffinity?

```go
switch {
  case mungedPodSpec.Affinity == nil && oldNodeAffinity == nil:
    // already effectively nil, no change needed
  case mungedPodSpec.Affinity == nil && oldNodeAffinity != nil
    mungedPodSpec.Affinity = &core.Affinity{NodeAffinity: oldNodeAffinity} // +k8s:verify-mutation:reason=clone
  default:
    mungedPodSpec.Affinity.NodeAffinity = oldNodeAffinity // +k8s:verify-mutation:reason=clone
}
```

danielvegamyhre replied:
I agree this makes the logic more clear - updated.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7297,3 +7329,68 @@ func validatePvNodeAffinity(newPvNodeAffinity, oldPvNodeAffinity *core.VolumeNod
+	// to matchExpressions or fieldExpressions are allowed.
+	for i := range oldTerms {
+		if !validateNodeSelectorTermHasOnlyAdditions(newTerms[i], oldTerms[i]) {
+			allErrs = append(allErrs, field.Invalid(nodeAffinityPath.Child("requiredDuringSchedulingIgnoredDuringExecution").Child("nodeSelectorTerms"), fmt.Sprintf("\nOLD: %v\nNEW: %v\n", oldTerms[i], newTerms[i]), "only additions are allowed (no mutations or deletions)"))
```
does this fmt.Sprint result in a readable error message? what does it output? the newlines in the message don't seem right. I'd probably do this:

```suggestion
			allErrs = append(allErrs, field.Invalid(nodeAffinityPath.Child("requiredDuringSchedulingIgnoredDuringExecution").Child("nodeSelectorTerms").Index(i), newTerms[i], "only additions are allowed (no mutations or deletions)"))
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7297,3 +7329,68 @@ func validatePvNodeAffinity(newPvNodeAffinity, oldPvNodeAffinity *core.VolumeNod
+		if len(newTerm.MatchFields) < l {
+			return false
+		}
+		if !reflect.DeepEqual(newTerm.MatchFields[:l], oldTerm.MatchFields) {
```
```suggestion
		if !apiequality.Semantic.DeepEqual(newTerm.MatchFields[:l], oldTerm.MatchFields) {
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7297,3 +7329,68 @@ func validatePvNodeAffinity(newPvNodeAffinity, oldPvNodeAffinity *core.VolumeNod
+		if len(newTerm.MatchExpressions) < l {
+			return false
+		}
+		if !reflect.DeepEqual(newTerm.MatchExpressions[:l], oldTerm.MatchExpressions) {
```
```suggestion
		if !apiequality.Semantic.DeepEqual(newTerm.MatchExpressions[:l], oldTerm.MatchExpressions) {
```

danielvegamyhre replied:
Updated. Just curious for my own learning, why does memory equality vs semantic equality make a difference here? Are there cases where two MatchExpressions may have different values but be semantically equivalent?

liggitt replied:
Semantic equality treats null and empty slices as equal. The distinction between null and [] is lost when round-tripping through protobuf storage anyway, but is preserved in JSON, so a JSON client setting `[]` in a request would get squashed to `null` when written to and read from etcd.

liggitt commented:
(integration flake was fixed by https://github.com/kubernetes/kubernetes/pull/116393)

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7297,3 +7331,68 @@ func validatePvNodeAffinity(newPvNodeAffinity, oldPvNodeAffinity *core.VolumeNod
+	return allErrs
+}
+
+func validateNodeSelectorTermHasOnlyAdditions(newTerm, oldTerm core.NodeSelectorTerm) bool {
```
@ahg-g am I reading the docs and implementation correctly that an empty NodeSelectorTerm (no match expressions and no match fields) matches nothing?

https://github.com/kubernetes/kubernetes/blob/20e44913853fba580997cacad594e5274613bd2d/pkg/apis/core/types.go#L2641-L2645

If so, that means we need something like this at the top of the function to avoid turning a term that selects nothing into a term that selects something:

```go
if len(oldTerm.MatchExpressions) == 0 && len(oldTerm.MatchFields) == 0 {
  if len(newTerm.MatchExpressions) > 0 || len(newTerm.MatchFields) > 0 {
    return false
  }
}
```

ahg-g replied:
Correct, wow, so much nuance!

danielvegamyhre replied:
Updated accordingly and added a unit test for this case: `empty NodeSelectorTerm (selects nothing) cannot become populated (selects something)`

liggitt commented:
/lgtm
/approve
Thanks for all the iterations, the test coverage on this is really good

/hold
Squash down to a single commit and fixup the commit message, then @ahg-g can re-lgtm

liggitt commented:
/triage accepted
/priority important-soon

ahg-g commented:
Thanks @liggitt for the detailed review!

ahg-g commented:
/lgtm
/hold cancel

danielvegamyhre commented:
Squashed commits and updated the commit message
